docs - reorg pxf content, add multi-server, objstore content

lisakowen opened the pull request:
in this PR:
- un-nested the main pxf section subnav (1 less click!)
- new pages for some existing topics
- reorganized the left navigation pane:
    - new "Administering PXF" topic that consolidates admin-related config info and procedures
    - new high level topic for Hadoop connectors; moved existing content for each supported hadoop data format to its own page
- added a new high level topic for object store connectors
    - new content and page for each data format supported by obj stores
- add parquet write (experimental) content

this PR does NOT currently include:
- docs for AvroSequenceFile data format (just a mention of support)
- PXF SDK doc updates - the topic and content have been temporarily removed

doc review site links:
- gpdb pxf protocol page: http://docs-lisa-pxf-try2.cfapps.io/600/admin_guide/external/g-pxf-protocol.html
- gpdb pxf landing page:  http://docs-lisa-pxf-try2.cfapps.io/600/admin_guide/external/pxf-overview.html
- pxf section: http://docs-lisa-pxf-try2.cfapps.io/600/pxf/overview_pxf.html

kongyew commented:
Suggestion for left navigation, so it is easy to use.
Reason 
- Typical use case: a new gpadmin will focus on PXF installation and configuration
- If PXF is already installed, gpadmin can proceed to choose "Upgrading PXF"
- "About the PXF installation and Configuration directories" can be categorized under "Configuring PXF".

Example:
Administering PXF:
- Configuring PXF
-- about PXF installation and Configuration directories
- Upgrading PXF
....

kongyew commented:
Changes:
"Reading and Writing SequenceFileData" left navigation points to the page with this title "Reading and Writing HDFS SequenceFile Data" .  Does it make sense to simplify the title to use "Reading and Write SequenceFile" since the context is obvious and direct.

Similarly, this changes can be apply to other left navigation items under "Accessing Hadoop with PXF"

kongyew commented:
Updates to "Accessing Azure, Google Cloud Storage, Minio, and S3 Object Stores with PXF" table..

The table shows profiles for each object storage .  Can we modify the profile table to so that it shows the format "S3 or Minio" ?

lisakowen commented:
thank you for reviewing, @denalex .  i have incorporated your requested changes.

dyozie commented on gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb:
```diff
@@ -125,15 +125,15 @@ In this procedure, you name and add a PXF server configuration in the `$PXF_CONF
 3. For each object store connector that you configure:
 
-    1. Determine a name for the server. You will provide the name to end users to whom you give permission to create external tables referencing files in the object store.
+    1. Determine a name for the server. You will provide the name to end users that need to referenc files in the object store.
```
Typo: referenc

lisakowen replied:
doh...  good catch!  i will correct that.

lisakowen commented:
@kongyew - i moved install/cfg dirs and upgrade section as you requested.  also updated the table column heading to include minio.

re: removing the "HDFS" and "Object Store" qualifiers from the page titles...  the hdfs and object store topics for a specific data format have the same title in the subnav (left pane).  i think that qualifying the name in the title of the content page provides the user an indicator of where they are in the docs.  our subnav doesn't always do a good job of that.  so i am leaning towards keeping the qualifiers in the page title.  (i will verify that we are doing this consistently.)

dyozie commented on gpdb-doc/markdown/pxf/access_objstore.html.md.erb:
````diff
@@ -56,8 +56,27 @@ The PXF connectors to Azure, Google Cloud Storage, Minio, and S3 expose the foll
-```
+</pre>
+
+## <a id="s3_override"></a>Overriding S3 Server Configuration
````
Overriding -> Overriding the

dyozie commented on gpdb-doc/markdown/pxf/access_objstore.html.md.erb:
```diff
@@ -56,8 +56,27 @@ The PXF connectors to Azure, Google Cloud Storage, Minio, and S3 expose the foll
+
+## <a id="s3_override"></a>Overriding S3 Server Configuration
+
+If you are accessing the S3 object store, you can override an S3 server configuration by directly specifying the S3 access ID and secret key via these custom options in the `CREATE EXTERNAL TABLE` `LOCATION` clause:
```
I think this would read better by converting some an's to the's and vice versa:

"If you are accessing **an** S3 object store, you can override **the** S3 server configuration by directly specifying the S3 access ID and secret key via ..."

Also, would it be clearer to say "override the _default_ S3 server configuration" ?

dyozie commented on gpdb-doc/markdown/pxf/access_objstore.html.md.erb:
```diff
@@ -56,8 +56,27 @@ The PXF connectors to Azure, Google Cloud Storage, Minio, and S3 expose the foll
+  LOCATION ('pxf://S3_BUCKET/dir/file.txt?PROFILE=s3:text&SERVER=s3srvcfg<b>&accesskey=YOURKEY&secretkey=YOURSECRET</b>')
+FORMAT 'TEXT' (delimiter=E',');</pre>
+
+<div class="note warning">Credentials that you provide in this manner are sent clear text, PXF does not encrypt them. Do not use this method of passing credentials in a production environment.</div>
```
sent clear -> sent in clear

dyozie commented on gpdb-doc/markdown/pxf/access_objstore.html.md.erb:
```diff
@@ -56,8 +56,27 @@ The PXF connectors to Azure, Google Cloud Storage, Minio, and S3 expose the foll
+
+<div class="note warning">Credentials that you provide in this manner are sent clear text, PXF does not encrypt them. Do not use this method of passing credentials in a production environment.</div>
+
+PXF does not support overriding Azure, Google Cloud Storage, and Minio server configuration at this time.
```
and Minio server configuration -> or Minio server configurations

dyozie commented on gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb:
```diff
@@ -43,69 +41,80 @@ For example, the contents of the `s3-site.xml` template file follow:
-7. Publish the PXF server names to your Greenplum Database end users as appropriate. The end user will specify the server name in the `CREATE EXTERNAL TABLE` `LOCATION` clause `SERVER` option to access the object store.
+7. Publish the PXF server names to your Greenplum Database end users as appropriate.
+
+The Greenplum Database user will specify the server name in the `CREATE EXTERNAL TABLE` `LOCATION` clause `SERVER` option to access the object store. For example:
```
will specify -> specifies

dyozie commented on gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb:
```diff
@@ -43,69 +41,80 @@ For example, the contents of the `s3-site.xml` template file follow:
+| Property       | Description                                | Value |
+|----------------|--------------------------------------------|-------|
+| fs.s3a.access.key | The AWS account access key ID. | Your access key. |
+| fs.s3a.secret.key | The secret key associated with the AWS access key ID. | Your secred key. |
```
Typo: secred

dyozie commented on gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb:
```diff
@@ -43,69 +41,80 @@ For example, the contents of the `s3-site.xml` template file follow:
 
-<div class="note warning">The PXF S3 object store connector also supports specifying the S3 access ID and secret key via custom options in the CREATE EXTERNAL TABLE command LOCATION clause. As with the credentials that you configure in the s3-site.xml file, these keys are clear text, PXF does not encrypt them. Do not use this method of passing credentials in a production environment.</div>
+<div class="note warning">You can override an S3 server configuration by directly specifying the S3 access ID and secret key via custom options in the CREATE EXTERNAL TABLE command LOCATION clause. For example:<br>
+<pre>CREATE EXTERNAL TABLE pxf_ext_tbl(name text, orders int)
```
sent clear -> sent in clear

dyozie commented on gpdb-doc/markdown/pxf/objstore_cfg.html.md.erb:
```diff
@@ -169,13 +178,11 @@ In this procedure, you name and add a PXF server configuration in the `$PXF_CONF
+## <a id="client-cfg-update"></a>Adding or Updating Object Store Configuration
 
-If you add or update object store server configuration while the PXF service is running, you must re-sync the PXF configuration to your Greenplum Database cluster and restart PXF on each segment host in the cluster. For example:
+If you add or update object store server configuration on the Greenplum Database master host, you must re-sync the PXF configuration to the Greenplum Database cluster:
```
update object -> update the object

dyozie commented on gpdb-doc/markdown/pxf/hdfs_parquet.html.md.erb:
```diff
@@ -21,50 +21,35 @@ specific language governing permissions and limitations
 -->
 
-Use the PXF HDFS connector to read and write parquet-format data. This section decribes how to read and write HDFS files that are stored in Parquet format, including how to create and query an external table that references files in the HDFS data store.
+Use the PXF HDFS connector to read and write parquet-format data. This section decribes how to read and write HDFS files that are stored in Parquet format, including how to create, query, and insert into an external table that references files in the HDFS data store.
```
Just a note for future edits -> Parquet isn't consistently capitalized throughout.  Fine to ignore this for the current release.

dyozie commented on gpdb-doc/markdown/pxf/client_instcfg.html.md.erb:
```diff
@@ -8,9 +8,9 @@ PXF is compatible with Cloudera, Hortonworks Data Platform, MapR, and generic Ap
+Configuring PXF Hadoop connectors involves copying configuration files from your Hadoop cluster to the Greenplum Database master host. If you are using the MapR Hadoop distribution, you must also copy certain JAR files to the master host. Before you configure the PXF Hadoop connectors, ensure that you can copy files from hosts in your Hadoop cluster to the Greenplum Database master.
 
-In this procedure, you copy Hadoop configuration files to the `$PXF_CONF/servers/default` directory on the Greenplum Database master host and synchronize the configuration to the segment hosts. PXF creates the`$PXF_CONF/servers/default` directory when you run `pxf cluster init`.
+In this procedure, you copy Hadoop configuration files to the `$PXF_CONF/servers/default` directory on the Greenplum Database master host. You may also copy libraries to `$PXF_CONF/lib. You then synchronize the PXF configuration on the master host to the segment hosts. (PXF creates the`$PXF_CONF/*` directories when you run `pxf cluster init`.)
```
Seems like second sentence should read something like: "You may also copy libraries to $PXF_CONF/lib for MapR support."  ?

lisakowen commented:
thank you for reviewing, @frankgh . i qualified the MapR config step with the requested note.
